picky-krb/src/crypto/diffie_hellman.rs

```diff
@@ -114,25 +119,43 @@ pub fn generate_key(
     nonce: Option<DhNonce>,
     cipher: &dyn Cipher,
 ) -> DiffieHellmanResult<Vec<u8>> {
-    let dh_shared_secret = generate_dh_shared_secret(public_key, private_key, modulus);
-
-    generate_key_from_shared_secret(&dh_shared_secret, nonce, cipher)
+    let dh_shared_secret = generate_dh_shared_secret(public_key, private_key, modulus)?;
+    generate_key_from_shared_secret(dh_shared_secret, nonce, cipher)
 }
 
 /// [Key and Parameter Requirements](https://www.rfc-editor.org/rfc/rfc2631#section-2.2)
 /// X9.42 requires that the private key x be in the interval [2, (q - 2)]
-pub fn generate_private_key<R: RngCore + CryptoRng>(q: &[u8], rng: &mut R) -> Vec<u8> {
-    let q = BigUint::from_bytes_be(q);
-
-    rng.gen_range(BigUint::from_bytes_be(&[2])..(q - 1_u32)).to_bytes_be()
+pub fn generate_private_key<R: TryCryptoRng>(q: &[u8], rng: &mut R) -> Result<Vec<u8>, R::Error> {
+    let q = BoxedUint::from_be_slice_vartime(q);
+    let low_bound = BoxedUint::from_be_slice_vartime(&[2]);
+    let high_bound = q - 1_u32;
+
+    let min_bits = low_bound.bits();
+    let max_bits = high_bound.bits();
+    loop {
+        let bit_length = rng.try_next_u32()? % (max_bits - min_bits) + min_bits;
+        let x = BoxedUint::random_bits(rng, bit_length);
+
+        if (&low_bound..&high_bound).contains(&&x) {
+            return Ok(x.to_be_bytes().into_vec());
+        }
+    }
 }
 
 /// [Key and Parameter Requirements](https://www.rfc-editor.org/rfc/rfc2631#section-2.2)
 /// y is then computed by calculating g^x mod p.
-pub fn compute_public_key(private_key: &[u8], modulus: &[u8], base: &[u8]) -> Vec<u8> {
-    let x = BigUint::from_bytes_be(private_key);
-    let g = BigUint::from_bytes_be(base);
-    let p = BigUint::from_bytes_be(modulus);
+pub fn compute_public_key(private_key: &[u8], modulus: &[u8], base: &[u8]) -> DiffieHellmanResult<Vec<u8>> {
+    generate_dh_shared_secret(base, private_key, modulus)
```

Comment:
The function body looked the same as `generate_dh_shared_secret`, so I added direct use of `generate_dh_shared_secret`.